### PR TITLE
feat(semantic_extract): schema-guided validation — SchemaValidator + ExtractionSchema (PR 1/3)

### PR DESCRIPTION
**Pull request by Besokus**
## Description

Part of #1510 — the first of the focused PRs agreed on the issue. Scope here is exactly the slice @KaifAhmad1 and @pkupt called a good first PR: **`SchemaValidator` + the `ExtractionSchema` view**. `bootstrap_schema` and schema-conditioned prompting are separate follow-ups.

This adds a deterministic, ontology-based validator — the core of ontology-based information extraction (OBIE; Wimalasuriya & Dou, 2010). No LLM, no new runtime dependencies.

Per @KaifAhmad1's interface feedback: `SchemaValidator` is a **separate class** that mirrors `ExtractionValidator` exactly — same two methods `validate_entities()` / `validate_relations()`, batch-aware, each returning the **same `ValidationResult`** type (imported, not re-defined; violations → `errors`, counts → `metrics`, `score` = conformance ratio). So the two validators run back-to-back on orthogonal axes (confidence vs. schema conformance).

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

## Related Issues

Part of #1510 (first of three focused PRs; does not fully close the issue on its own).

## Changes Made

- New `ExtractionSchema` (`semantica.semantic_extract`, lazy export): a read-only view over a domain ontology — allowed concepts + predicates with optional `domain`/`range`. Reuses the existing OWL representation rather than a parallel type. Build one from a `generate_ontology()`-style dict (`from_ontology`) or an OWL/Turtle file/string (`from_owl`, via the existing `rdflib` dependency). Empty `domain`/`range` = unconstrained, matching OWL.
- New `SchemaValidator` (`semantica.semantic_extract`, lazy export): entity labels must be schema concepts; relation predicates must be in the schema and satisfy `domain`/`range`. `filter_by_schema()` / `filter_relations_by_schema()` return the conforming subset (mirroring `filter_by_confidence`).
- Registered both in `semantic_extract/__init__.py` (lazy export + `__all__`); added a `CHANGELOG.md` entry.

## Testing

- [x] Tested locally
- [x] Added tests for new functionality
- [ ] Package builds successfully (`python -m build`)

`tests/semantic_extract/test_schema_validator.py` — 12 tests: schema parsing (`from_ontology`, `from_owl`), `domain`/`range` + wildcard logic, entity/relation validation (conforming, out-of-vocabulary, unknown predicate, domain/range violation, empty, batch), filtering, and back-to-back composition with `ExtractionValidator`.

### Test Commands

```bash
pytest tests/semantic_extract/test_schema_validator.py   # 12 passed
black --check semantica/semantic_extract/ tests/semantic_extract/test_schema_validator.py
isort --check-only semantica/semantic_extract/ tests/semantic_extract/test_schema_validator.py

Documentation

- [ ] Updated relevant documentation
- [ ] Added code examples if applicable
- [ ] Updated API reference if adding new APIs
- [ ] Updated cookbook if adding new examples
- [x] No documentation changes needed

A cookbook notebook demonstrating the full loop will accompany the follow-up PRs (bootstrap_schema / schema-conditioned prompting).

Breaking Changes

Breaking Changes: No

Additive only — new modules and lazy exports; no existing behavior changes.

Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] Package builds successfully

Additional Notes

Follow-ups on #1510: PR2 — bootstrap_schema as a thin wrapper over OntologyGenerator (reusing its concept/property + domain/range induction and TTL export) plus a min_occurrences pre-filter, per @KaifAhmad1 (@pkupt volunteered); PR3 — optional schema-conditioned prompting. No new runtime dependencies in this PR.

### CODE REVIEW

File: semantica/semantic_extract/schema.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    def from_ontology(cls, ontology: Mapping[str, Any]) -> "ExtractionSchema":
+        """Build a schema from a ``generate_ontology``-style dict.
+
+        Reads ``ontology["classes"]`` (each carrying a ``name`` / ``label``) as
+        concepts and ``ontology["properties"]`` (each carrying a ``name`` and
+        optional ``domain`` / ``range``) as predicates. Missing ``domain`` /
+        ``range`` means unconstrained; unrecognised keys are ignored.
+        """
+        concepts: Set[str] = set()
+        for c in ontology.get("classes", []) or []:
```
[the remark this answers is not in the document]
Comment (Besokus) [reply semantica/semantic_extract/schema.py:129]: Good catch — this is real. semantica/ingest/ontology_ingestor.py returns an OntologyData whose ontology dict is held in .data, and from_ontology would have raised AttributeError on it. Fixed in 4bdd4948: from_ontology now unwraps an OntologyData-like object (.data) before parsing, so both the generate_ontology() dict and the ingestion path work. Added a test — thanks!

File: semantica/semantic_extract/schema.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            name = (c.get("name") or c.get("label")) if isinstance(c, Mapping) else c
+            if name:
+                concepts.add(str(name))
```
[the remark this answers is not in the document]
Comment (Besokus) [reply semantica/semantic_extract/schema.py:132]: By design. Names are matched exactly, so the schema vocabulary and the extraction labels must share a normalization convention; this is now documented on ExtractionSchema (OntologyGenerator normalizes to PascalCase / camelCase). I chose documentation over silent case-folding, which would mask real mismatches. Resolving.

File: semantica/semantic_extract/schema.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        """
+        if subject_label not in self.concepts or object_label not in self.concepts:
+            return False
+        pred = self.predicates.get(predicate)
+        if pred is None:
+            return False
+        if pred.domain and subject_label not in pred.domain:
+            return False
+        if pred.range and object_label not in pred.range:
+            return False
```
[the remark this answers is not in the document]
Comment (Besokus) [reply semantica/semantic_extract/schema.py:233]: Documented as a known limitation / possible follow-up on allows_relation: membership is exact and subClassOf hierarchies are not traversed. Carrying the class hierarchy into the schema is out of scope for this first PR. Resolving.

**Review by ZohaibHassan16 — APPROVED** [review]
LGTM. Addressed all the feedback cleanly, edge cases and `owl:Thing` handling look great, and all 17 tests are green. Ready to merge!
